File: model-api/src/main/java/com/mgmtp/a12/print/model/api/walker/model/PrintModelVisitor.java
/*
 * SPDX-License-Identifier: EUPL-1.2 OR LicenseRef-commercial
 *
 * Copyright (c) 2012-2026 mgm technology partners GmbH
 *
 * Dual License
 * ------------
 * This source file is part of the mgm A12 Platform and available under
 * a choice of two different licenses:
 *
 * 1. Open-Source License - EUPL v1.2
 *    You may redistribute and/or modify this file under the terms of the
 *    European Union Public License, version 1.2 - see https://eupl.eu/.
 *
 * 2. Commercial License
 *    Alternatively, you may obtain a commercial license from
 *    mgm technology partners GmbH, that permits use of this software
 *    under different terms (including support and maintenance services).
 *
 *    Please contact a12-license@mgm-tp.com for more information.
 *
 * You must select and comply with exactly one of the above license options.
 *
 * Warranty Disclaimer (applies to either option)
 * ----------------------------------------------
 * THIS SOFTWARE IS PROVIDED "AS IS" AND WITHOUT WARRANTY OF ANY KIND,
 * WHETHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NON-INFRINGEMENT, EXCEPT WHERE SUCH DISCLAIMERS ARE HELD TO BE
 * LEGALLY INVALID. SEE THE RESPECTIVE LICENSE TEXT FOR DETAILS.
 */
package com.mgmtp.a12.print.model.api.walker.model;

import com.mgmtp.a12.print.model.api.exceptions.*;
import com.mgmtp.a12.print.model.api.model.PrintModel;
import com.mgmtp.a12.print.model.api.model.container.BaseReferenceContainer;
import com.mgmtp.a12.print.model.api.model.element.PrintModelElement;
import com.mgmtp.a12.print.model.api.model.element.type.area.Area;
import com.mgmtp.a12.print.model.api.model.element.type.boundingBox.BoundingBox;
import com.mgmtp.a12.print.model.api.model.element.type.calculation.Calculation;
import com.mgmtp.a12.print.model.api.model.element.type.chart.barChart.BarChart;
import com.mgmtp.a12.print.model.api.model.element.type.chart.lineChart.LineChart;
import com.mgmtp.a12.print.model.api.model.element.type.chart.pieChart.PieChart;
import com.mgmtp.a12.print.model.api.model.element.type.expression.Expression;
import com.mgmtp.a12.print.model.api.model.element.type.field.Field;
import com.mgmtp.a12.print.model.api.model.element.type.horizontalLine.HorizontalLine;
import com.mgmtp.a12.print.model.api.model.element.type.image.Image;
import com.mgmtp.a12.print.model.api.model.element.type.listing.Listing;
import com.mgmtp.a12.print.model.api.model.element.type.override.OverrideElement;
import com.mgmtp.a12.print.model.api.model.element.type.pageNumber.PageNumber;
import com.mgmtp.a12.print.model.api.model.element.type.pageNumber.PageNumberTotal;
import com.mgmtp.a12.print.model.api.model.element.type.switchCase.Switch;
import com.mgmtp.a12.print.model.api.model.element.type.table.Table;
import com.mgmtp.a12.print.model.api.model.element.type.tableLayout.TableLayout;
import com.mgmtp.a12.print.model.api.model.element.type.text.TextElement;
import com.mgmtp.a12.print.model.api.model.general.Metadata;
import com.mgmtp.a12.print.model.api.model.reference.ElementReference;
import com.mgmtp.a12.print.model.api.model.reference.PlaceableReference;
import com.mgmtp.a12.print.model.api.model.reference.TableColumnReference;
import com.mgmtp.a12.print.model.api.model.reference.TableLayoutCellReference;
import com.mgmtp.a12.print.model.api.model.section.ModelSection;
import com.mgmtp.a12.print.model.api.model.segment.ModelSegment;
import com.mgmtp.a12.print.model.api.model.watermark.Watermark;
import com.mgmtp.a12.print.model.api.walker.DescendCommand;
import com.mgmtp.a12.print.model.api.walker.TraversalCommand;

/**
 * Provides an interface to execute a given function on every print model entity.
 * Uses {@link DescendCommand} and {@link TraversalCommand} return values to control the navigation of the corresponding {@link PrintModelWalker}.
 */
public interface PrintModelVisitor {

	default void beforeVisitElement(final PrintModelElement element, final PrintModelPath path) {

	}

	default void afterVisitElement(final PrintModelElement element, final PrintModelPath path) {
	}

	default DescendCommand descendPrintModel(final PrintModel printModel) {
		return DescendCommand.DESCEND_FIRST;
	}

	default DescendCommand descendDINTemplate(
		final ModelSegment dinTemplate,
		final PrintModelPath path
	) {
		return DescendCommand.DESCEND_FIRST;
	}

	default DescendCommand descendContainer(
		final BaseReferenceContainer<? extends ElementReference> container,
		final PrintModelPath path,
		final int index
	) {
		return DescendCommand.DESCEND_FIRST;
	}

	default TraversalCommand visitPrintModel(final PrintModel printModel) {
		return TraversalCommand.CONTINUE;
	}

	default TraversalCommand visitDINTemplate(
		final ModelSegment dinTemplate,
		final PrintModelPath path
	) {
		return TraversalCommand.CONTINUE;
	}

	default TraversalCommand visitUnresolvedDinTemplate(
		final String dinTemplateReference
	) {
		throw new UnresolvedDinTemplateException(dinTemplateReference);
	}

	default TraversalCommand visitContainer(
		final BaseReferenceContainer<? extends ElementReference> container,
		final PrintModelPath path
	) {
		if (container instanceof PrintModelElement) {
			return this.visitElement((PrintModelElement) container, path);
		}
		return this.visitReferenceContainer(container, path);
	}

	default TraversalCommand visitReferenceContainer(
		final BaseReferenceContainer<? extends ElementReference> container,
		final PrintModelPath path
	) {
		if (container instanceof ModelSegment) {
			return visitSegment((ModelSegment) container, path);
		}
		if (container instanceof ModelSection) {
			return visitSection((ModelSection) container, path);
		}
		if (container instanceof Watermark) {
			return visitWatermark((Watermark) container, path);
		}
		return visitUnresolvedReferenceContainer(container, path);
	}

	default TraversalCommand visitUnresolvedReferenceContainer(
		final BaseReferenceContainer<? extends ElementReference> container,
		final PrintModelPath path
	) {
		throw new UnresolvedReferenceContainerException(container);
	}

	default TraversalCommand visitSegment(final ModelSegment segment, final PrintModelPath path) {
		return TraversalCommand.CONTINUE;
	}

	default TraversalCommand visitSection(final ModelSection section, final PrintModelPath path) {
		return TraversalCommand.CONTINUE;
	}

	default TraversalCommand visitWatermark(final Watermark watermark, final PrintModelPath path) {
		return TraversalCommand.CONTINUE;
	}

	default TraversalCommand visitMetadata(final Metadata metadata, final PrintModelPath path) {
		return TraversalCommand.CONTINUE;
	}

	default TraversalCommand visitReference(
		final ElementReference reference,
		final PrintModelPath path,
		final int index
	) {
		if (reference instanceof PlaceableReference) {
			return visitPlaceableReference((PlaceableReference) reference, path, index);
		}
		if (reference instanceof TableLayoutCellReference) {
			return visitTableLayoutCellReference((TableLayoutCellReference) reference, path, index);
		}
		if (reference instanceof TableColumnReference) {
			return visitTableColumnReference((TableColumnReference) reference, path, index);
		}

		return TraversalCommand.CONTINUE;
	}

	default TraversalCommand visitPlaceableReference(
		final PlaceableReference reference,
		final PrintModelPath path,
		final int index
	) {
		return TraversalCommand.CONTINUE;
	}

	default TraversalCommand visitTableLayoutCellReference(
		final TableLayoutCellReference reference,
		final PrintModelPath path,
		final int index
	) {
		return TraversalCommand.CONTINUE;
	}

	default TraversalCommand visitTableColumnReference(
		final TableColumnReference reference,
		final PrintModelPath path,
		final int index
	) {
		return TraversalCommand.CONTINUE;
	}

	default TraversalCommand visitElement(final PrintModelElement element, final PrintModelPath path) {
		if (element instanceof TextElement) {
			return visitText((TextElement) element, path);
		} else if (element instanceof Expression) {
			return visitExpression((Expression) element, path);
		} else if (element instanceof HorizontalLine) {
			return visitHorizontalLine((HorizontalLine) element, path);
		} else if (element instanceof Image) {
			return visitImage((Image) element, path);
		} else if (element instanceof Listing) {
			return visitListing((Listing) element, path);
		} else if (element instanceof Table) {
			return visitTable((Table) element, path);
		} else if (element instanceof TableLayout) {
			return visitTableLayout((TableLayout) element, path);
		} else if (element instanceof BarChart) {
			return visitBarChart((BarChart) element, path);
		} else if (element instanceof LineChart) {
			return visitLineChart((LineChart) element, path);
		} else if (element instanceof PieChart) {
			return visitPieChart((PieChart) element, path);
		} else if (element instanceof Field) {
			return visitField((Field) element, path);
		} else if (element instanceof Calculation) {
			return visitCalculation((Calculation) element, path);
		} else if (element instanceof PageNumber) {
			return visitPageNumber((PageNumber) element, path);
		} else if (element instanceof PageNumberTotal) {
			return visitPageNumberTotal((PageNumberTotal) element, path);
		} else if (element instanceof BoundingBox) {
			return visitBoundingBox((BoundingBox) element, path);
		} else if (element instanceof Area) {
			return visitArea((Area) element, path);
		} else if (element instanceof Switch) {
			return visitSwitch((Switch) element, path);
		} else if (element instanceof OverrideElement) {
			return visitOverrideElement((OverrideElement) element, path);
		}

		return visitUnknownElement(element, path);
	}

	default TraversalCommand visitUnresolvedElement(
		final ElementReference reference,
		final PrintModelPath path,
		final int index
	) {
		throw new UnresolvedElementException(reference, index);
	}

	default TraversalCommand visitUnresolvedSegment(
		final String segmentId,
		final int index
	) {
		throw new UnresolvedSegmentException(segmentId, index);
	}

	default TraversalCommand visitUnresolvedSection(
		final String sectionId,
		final int index
	) {
		throw new UnresolvedSectionException(sectionId, index);
	}

	default TraversalCommand visitUnresolvedWatermark(
		final String watermarkId,
		final int index
	) {
		throw new UnresolvedWatermarkException(watermarkId, index);
	}

	default TraversalCommand visitUnknownElement(
		final PrintModelElement element,
		final PrintModelPath path
	) {
		return TraversalCommand.CONTINUE;
	}

	default TraversalCommand defaultVisitElement(final PrintModelElement element, final PrintModelPath path) {
		return TraversalCommand.CONTINUE;
	}

	default TraversalCommand visitExpression(final Expression expression, final PrintModelPath path) {
		return defaultVisitElement(expression, path);
	}

	default TraversalCommand visitListing(final Listing listing, final PrintModelPath path) {
		return defaultVisitElement(listing, path);
	}

	default TraversalCommand visitTable(final Table table, final PrintModelPath path) {
		return defaultVisitElement(table, path);
	}

	default TraversalCommand visitTableLayout(final TableLayout staticTableLayout, final PrintModelPath path) {
		return defaultVisitElement(staticTableLayout, path);
	}

	default TraversalCommand visitBarChart(final BarChart barChart, final PrintModelPath path) {
		return defaultVisitElement(barChart, path);
	}

	default TraversalCommand visitLineChart(final LineChart lineChart, final PrintModelPath path) {
		return defaultVisitElement(lineChart, path);
	}

	default TraversalCommand visitPieChart(final PieChart pieChart, final PrintModelPath path) {
		return defaultVisitElement(pieChart, path);
	}

	default TraversalCommand visitImage(final Image image, final PrintModelPath path) {
		return defaultVisitElement(image, path);
	}

	default TraversalCommand visitHorizontalLine(final HorizontalLine horizontalLine, final PrintModelPath path) {
		return defaultVisitElement(horizontalLine, path);
	}

	default TraversalCommand visitPageNumber(final PageNumber pageNumber, final PrintModelPath path) {
		return defaultVisitElement(pageNumber, path);
	}

	default TraversalCommand visitPageNumberTotal(final PageNumberTotal pageNumberTotal, final PrintModelPath path) {
		return defaultVisitElement(pageNumberTotal, path);
	}

	default TraversalCommand visitCalculation(final Calculation calculation, final PrintModelPath path) {
		return defaultVisitElement(calculation, path);
	}

	default TraversalCommand visitField(final Field field, final PrintModelPath path) {
		return defaultVisitElement(field, path);
	}

	default TraversalCommand visitText(final TextElement text, final PrintModelPath path) {
		return defaultVisitElement(text, path);
	}

	default TraversalCommand visitBoundingBox(final BoundingBox box, final PrintModelPath path) {
		return defaultVisitElement(box, path);
	}

	default TraversalCommand visitArea(final Area area, final PrintModelPath path) {
		return defaultVisitElement(area, path);
	}

	default TraversalCommand visitSwitch(final Switch switchElement, final PrintModelPath path) {
		return defaultVisitElement(switchElement, path);
	}

	default TraversalCommand visitOverrideElement(final OverrideElement overrideElement, final PrintModelPath path) {
		return defaultVisitElement(overrideElement, path);
	}

	default TraversalCommand visitOverriddenBoundingBox(
		final BoundingBox boundingBox,
		final PrintModelPath path,
		final OverrideElement overrideElement
	) {
		return TraversalCommand.CONTINUE;
	}
}

File: model-api/src/main/typescript/walker/print-model-visitor.ts
/*
 * SPDX-License-Identifier: EUPL-1.2 OR LicenseRef-commercial
 *
 * Copyright (c) 2012-2026 mgm technology partners GmbH
 *
 * Dual License
 * ------------
 * This source file is part of the mgm A12 Platform and available under
 * a choice of two different licenses:
 *
 * 1. Open-Source License - EUPL v1.2
 *    You may redistribute and/or modify this file under the terms of the
 *    European Union Public License, version 1.2 - see https://eupl.eu/.
 *
 * 2. Commercial License
 *    Alternatively, you may obtain a commercial license from
 *    mgm technology partners GmbH, that permits use of this software
 *    under different terms (including support and maintenance services).
 *
 *    Please contact a12-license@mgm-tp.com for more information.
 *
 * You must select and comply with exactly one of the above license options.
 *
 * Warranty Disclaimer (applies to either option)
 * ----------------------------------------------
 * THIS SOFTWARE IS PROVIDED "AS IS" AND WITHOUT WARRANTY OF ANY KIND,
 * WHETHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NON-INFRINGEMENT, EXCEPT WHERE SUCH DISCLAIMERS ARE HELD TO BE
 * LEGALLY INVALID. SEE THE RESPECTIVE LICENSE TEXT FOR DETAILS.
 */
// Disabling unused vars rule due to the need to maintain method signatures in abstract classes and their derivatives.
/* eslint-disable @typescript-eslint/no-unused-vars */
import {
	type AnyContainerElement,
	type AnyTopLevelContainerElement,
	Area,
	BarChart,
	BoundingBox,
	Calculation,
	Expression,
	Field,
	Image,
	isPlaceableReference,
	isSection,
	isSegment,
	isTableLayoutCellReference,
	isWatermark,
	Line,
	LineChart,
	Listing,
	Override,
	PageNumber,
	PageNumberTotal,
	type PartialReference,
	type PartialTableColumnReference,
	PieChart,
	type PlaceableReference,
	type PrintModel,
	PrintModelElement,
	type Reference,
	type Section,
	type Segment,
	Switch,
	Table,
	TableLayout,
	type TableLayoutCellReference,
	Text,
	type Watermark,
} from "../model/index.js";

import type { PrintModelTrace } from "./print-model-trace.js";
import type { PartialPrintModelTrace } from "./partial/index.js";

export class UnvisitedElement extends Error {
	constructor(element: PrintModelElement) {
		super(`Not exhaustive visitation, element: ${element.type} was visited by the default method.`);
		Object.setPrototypeOf(this, UnvisitedElement.prototype);
	}
}

export enum DescendCommand {
	NO_DESCEND,
	DESCEND_FIRST,
	ELEMENT_FIRST,
}

export enum TraversalCommand {
	STOP,
	HALT,
	CONTINUE,
}

export abstract class PrintModelVisitor {
	abstract beforeVisitElement(element: PrintModelElement, printModelTrace: PrintModelTrace): void;

	abstract afterVisitElement(element: PrintModelElement, printModelTrace: PrintModelTrace): void;

	descendPrintModel(printModel: PrintModel): DescendCommand {
		return DescendCommand.DESCEND_FIRST;
	}

	descendContainer(
		container: AnyContainerElement | PrintModelElement,
		printModelTrace: PrintModelTrace,
		index?: number | undefined
	): DescendCommand {
		return DescendCommand.DESCEND_FIRST;
	}

	descendDinTemplate(dinTemplate: Segment): DescendCommand {
		return DescendCommand.DESCEND_FIRST;
	}

	visitPrintModel(printModel: PrintModel): TraversalCommand {
		return TraversalCommand.CONTINUE;
	}

	visitContainer(
		container: AnyContainerElement | PrintModelElement,
		printModelTrace: PrintModelTrace
	): TraversalCommand {
		if (PrintModelElement.isInstance(container)) {
			return this.visitElement(container, printModelTrace);
		}
		return this.visitReferenceContainer(container, printModelTrace);
	}

	visitReferenceContainer(
		referenceContainer: AnyTopLevelContainerElement,
		printModelTrace: PrintModelTrace
	): TraversalCommand {
		if (isSegment(referenceContainer)) {
			return this.visitSegment(referenceContainer, printModelTrace);
		}
		if (isSection(referenceContainer)) {
			return this.visitSection(referenceContainer, printModelTrace);
		}
		if (isWatermark(referenceContainer)) {
			return this.visitWatermark(referenceContainer, printModelTrace);
		}
		return this.visitUnresolvedReferenceContainer(referenceContainer, printModelTrace);
	}

	visitUnresolvedReferenceContainer(
		referenceContainer: AnyContainerElement,
		printModelTrace: PrintModelTrace
	): TraversalCommand {
		return TraversalCommand.HALT;
	}

	visitSegment(segment: Segment, printModelTrace: PrintModelTrace): TraversalCommand {
		return TraversalCommand.CONTINUE;
	}

	visitSection(section: Section, printModelTrace: PrintModelTrace): TraversalCommand {
		return TraversalCommand.CONTINUE;
	}

	visitWatermark(watermark: Watermark, printModelTrace: PrintModelTrace): TraversalCommand {
		return TraversalCommand.CONTINUE;
	}

	visitReference(reference: Reference, printModelTrace: PrintModelTrace, index?: number): TraversalCommand {
		if (isPlaceableReference(reference)) {
			return this.visitPlaceableReference(reference, printModelTrace, index);
		}
		if (isTableLayoutCellReference(reference)) {
			return this.visitTableLayoutCellReference(reference, printModelTrace, index);
		}

		const parent = printModelTrace.getParent()?.parent;

		if (!parent || !PrintModelElement.isInstance(parent)) {
			return this.visitUnknownReference(reference, printModelTrace, index);
		}

		switch (true) {
			case Table.isInstance(parent):
				return this.visitTableColumnReference(reference, printModelTrace, index);
			case Text.isInstance(parent):
				return this.visitTextEntityReference(reference, printModelTrace, index);
			case Switch.isInstance(parent):
				return this.visitSwitchCaseReference(reference, printModelTrace, index);
			default:
				return this.visitUnknownReference(reference, printModelTrace, index);
		}
	}

	visitPlaceableReference(
		reference: PlaceableReference,
		printModelTrace: PrintModelTrace,
		index?: number
	): TraversalCommand {
		return TraversalCommand.CONTINUE;
	}

	visitTableLayoutCellReference(
		reference: TableLayoutCellReference,
		printModelTrace: PrintModelTrace,
		index?: number
	): TraversalCommand {
		return TraversalCommand.CONTINUE;
	}

	visitTableColumnReference(
		reference: PartialTableColumnReference,
		printModelTrace: PartialPrintModelTrace,
		index?: number
	): TraversalCommand {
		return TraversalCommand.CONTINUE;
	}

	visitTextEntityReference(
		reference: PartialReference,
		printModelTrace: PartialPrintModelTrace,
		index?: number
	): TraversalCommand {
		return TraversalCommand.CONTINUE;
	}

	visitSwitchCaseReference(
		reference: PartialReference,
		printModelTrace: PartialPrintModelTrace,
		index?: number
	): TraversalCommand {
		return TraversalCommand.CONTINUE;
	}

	visitUnknownReference(reference: PartialReference, printModelTrace: PartialPrintModelTrace, index?: number) {
		return TraversalCommand.HALT;
	}

	visitElement(element: PrintModelElement, printModelTrace: PrintModelTrace): TraversalCommand {
		if (Text.isInstance(element)) {
			return this.visitText(element, printModelTrace);
		}
		if (Table.isInstance(element)) {
			return this.visitTable(element, printModelTrace);
		}
		if (TableLayout.isInstance(element)) {
			return this.visitTableLayout(element, printModelTrace);
		}
		if (PieChart.isInstance(element)) {
			return this.visitPieChart(element, printModelTrace);
		}
		if (PageNumber.isInstance(element)) {
			return this.visitPageNumber(element, printModelTrace);
		}
		if (PageNumberTotal.isInstance(element)) {
			return this.visitPageNumberTotal(element, printModelTrace);
		}
		if (Listing.isInstance(element)) {
			return this.visitListing(element, printModelTrace);
		}
		if (Line.isInstance(element)) {
			return this.visitLine(element, printModelTrace);
		}
		if (LineChart.isInstance(element)) {
			return this.visitLineChart(element, printModelTrace);
		}
		if (Image.isInstance(element)) {
			return this.visitImage(element, printModelTrace);
		}
		if (Field.isInstance(element)) {
			return this.visitField(element, printModelTrace);
		}
		if (Expression.isInstance(element)) {
			return this.visitExpression(element, printModelTrace);
		}
		if (Calculation.isInstance(element)) {
			return this.visitCalculation(element, printModelTrace);
		}
		if (BarChart.isInstance(element)) {
			return this.visitBarChart(element, printModelTrace);
		}
		if (Switch.isInstance(element)) {
			return this.visitSwitch(element, printModelTrace);
		}
		if (Area.isInstance(element)) {
			return this.visitArea(element, printModelTrace);
		}
		if (BoundingBox.isInstance(element)) {
			return this.visitBoundingBox(element, printModelTrace);
		}
		if (Override.isInstance(element)) {
			return this.visitOverride(element, printModelTrace);
		}
		return this.visitUnknownElement(element, printModelTrace);
	}

	visitText(element: Text, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitTable(element: Table, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitTableLayout(element: TableLayout, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitPieChart(element: PieChart, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitPageNumber(element: PageNumber, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitPageNumberTotal(element: PageNumberTotal, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitListing(element: Listing, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitLine(element: Line, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitLineChart(element: LineChart, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitImage(element: Image, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitField(element: Field, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitExpression(element: Expression, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitCalculation(element: Calculation, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitBarChart(element: BarChart, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitBoundingBox(element: BoundingBox, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitArea(element: Area, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitOverride(element: Override, printModelTrace: PrintModelTrace): TraversalCommand {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitOverriddenBoundingBox(
		boundingBox: BoundingBox,
		override: Override,
		printModelTrace: PrintModelTrace
	): TraversalCommand {
		return TraversalCommand.CONTINUE;
	}

	visitSwitch(element: Switch, printModelTrace: PrintModelTrace) {
		return this.defaultVisitElement(element, printModelTrace);
	}

	visitDINTemplate(dinTemplate: Segment): TraversalCommand {
		return TraversalCommand.CONTINUE;
	}

	visitUnresolvedElement(reference: Reference, printModelTrace: PrintModelTrace, index?: number): TraversalCommand {
		return TraversalCommand.HALT;
	}

	visitUnresolvedSegment(id: string, index?: number): TraversalCommand {
		return TraversalCommand.HALT;
	}

	visitUnresolvedSection(id: string, index?: number): TraversalCommand {
		return TraversalCommand.HALT;
	}

	visitUnresolvedWatermark(id: string, index?: number): TraversalCommand {
		return TraversalCommand.HALT;
	}

	visitUnresolvedDinTemplate(templateId: string): TraversalCommand {
		return TraversalCommand.HALT;
	}

	visitUnknownElement(element: PrintModelElement, printModelTrace: PrintModelTrace): TraversalCommand {
		return TraversalCommand.HALT;
	}

	defaultVisitElement(element: PrintModelElement, printModelTrace: PrintModelTrace): TraversalCommand {
		return TraversalCommand.CONTINUE;
	}
}

export abstract class ExhaustivePrintModelVisitor extends PrintModelVisitor {
	defaultVisitElement(element: PrintModelElement): TraversalCommand {
		throw new UnvisitedElement(element);
	}
}
